### PR TITLE
[edward] Lion optimizer sweep vs AdamW

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Lion optimizer (Chen et al. 2023, "Symbolic Discovery of Optimization Algorithms") uses sign-based gradient updates with a much smaller memory footprint than AdamW, and has shown 1–2% improvements on image and language tasks over AdamW at the same LR budget. For DrivAerML CFD surrogates, the question is whether Lion's sign-based update finds a better loss basin than AdamW's adaptive moments — particularly for the wsy/wsz channels which have high signal-to-noise ratio.

Lion uses `update = beta1 * m + (1-beta1) * grad`, then `p -= lr * sign(update)`. The weight decay is applied as decoupled L2.

## Instructions

Base recipe: 4L/256d, no EMA, Fourier PE, T_max=30, surface_weight=2.0, 64 slices.

**Note**: Lion is sensitive to LR — typically optimal LR is 3–10× smaller than AdamW. Start with lr=1e-4 and lr=3e-4.

### Changes to train.py

1. Add `--optimizer` flag: choices=["adamw", "lion"], default "adamw".
2. If "lion": install `lion-pytorch` (pip) or implement inline:
   ```python
   class Lion(torch.optim.Optimizer):
       def __init__(self, params, lr=1e-4, betas=(0.9, 0.99), weight_decay=0.0):
           ...
       def step(self):
           for group in self.param_groups:
               for p in group['params']:
                   if p.grad is None: continue
                   grad = p.grad
                   m = self.state[p].get('m', torch.zeros_like(p))
                   update = group['betas'][0] * m + (1 - group['betas'][0]) * grad
                   p.data.mul_(1 - group['lr'] * group['weight_decay'])
                   p.data.add_(torch.sign(update), alpha=-group['lr'])
                   self.state[p]['m'] = group['betas'][1] * m + (1 - group['betas'][1]) * grad
   ```
3. Keep cosine annealing schedule (T_max=30) with Lion.

### Run A — Lion lr=1e-4

```bash
cd target/ && python train.py \
  --epochs 30 --lr 1e-4 --depth 4 --hidden-dim 256 --num-slices 64 \
  --surface-weight 2.0 --fourier-pe --no-ema \
  --train-surface-points 32768 --train-volume-points 32768 \
  --optimizer lion \
  --wandb-project DrivAerML --wandb-group edward-lion-sweep \
  --kill-thresholds "3000:val_primary/abupt_axis_mean_rel_l2_pct<=25"
```

### Run B — Lion lr=3e-4

```bash
cd target/ && python train.py \
  --epochs 30 --lr 3e-4 --depth 4 --hidden-dim 256 --num-slices 64 \
  --surface-weight 2.0 --fourier-pe --no-ema \
  --train-surface-points 32768 --train-volume-points 32768 \
  --optimizer lion \
  --wandb-project DrivAerML --wandb-group edward-lion-sweep \
  --kill-thresholds "3000:val_primary/abupt_axis_mean_rel_l2_pct<=25"
```

Run both; report val and test_primary metrics for each. If either Lion run beats AdamW baseline (abupt=7.209%), submit the result.

## Baseline

alphonse Wave 1 run `m9775k1v` (AdamW, lr=5e-4): abupt=7.209%, surf_p=4.802%, vol_p=4.166%✓, wsx=7.109%, wsy=9.100%←BINDING, wsz=10.869%←HARDEST

AB-UPT targets (all must be beaten simultaneously):
- surf_p: 3.82%
- vol_p: 6.08% ✓ (already beaten)
- wsx: 5.35%
- wsy: 3.65%
- wsz: 3.63%

### CODE REVIEW

**Comment by morganmcg1** [comment]
**ADVISOR check-in — Lion optimizer sweep**

Solid experiment. Lion has shown promise in vision transformers and the sign-based updates may help with the spiky wall-shear loss landscape. Key notes:

**Kill threshold is dead** — `3000:val_primary/abupt_axis_mean_rel_l2_pct<=25` fires at step 3000, before the first val at ~step 17,816. Update to `35000:val_primary/abupt_axis_mean_rel_l2_pct<20`.

**LR sensitivity** — Lion is typically 3-10× more sensitive to LR than AdamW. Your sweep of 1e-4 and 3e-4 is well-calibrated for this range. If both underperform baseline significantly by ep5, consider also trying 5e-5 as a rescue run (Lion can diverge at 3e-4 on some transformer variants).

**Weight decay** — With Lion, the effective weight decay is lr × wd, not just wd. If you're using the same wd=1e-2 as AdamW baseline, you may need to scale it down proportionally to the LR ratio (e.g., if Lion lr = 1/5 of AdamW lr, consider wd=5e-2 to maintain same regularization strength).

**Comparison baseline** — Reference edward's AdamW b1=0.95 run (, abupt=11.803%) as the AdamW variant to compare against, alongside the Wave 1 baseline. If Lion at lr=1e-4 beats AdamW b1=0.95 at the same LR scale, that's a clean signal.

**Report individual axis metrics** — especially wsy and wsz. Baseline wsy=9.100%, wsz=10.869% (target: 3.65%, 3.63%).

Baseline to beat: val abupt=7.2091% (PR #74, run m9775k1v).

**Comment by morganmcg1** [comment]
ADVISOR ESCALATION — second check-in.

Edward — this is your second advisor message. PR #217 (Lion optimizer sweep) has been in WIP for over 24 hours with no student response. Your pod shows 1/1 READY.

Please report:
1. Have you launched the Lion lr=1e-4 and lr=3e-4 runs?
2. If yes: W&B run IDs for both runs and current epoch
3. If not: implementation blockers or questions

Reminder: use the correct kill threshold (35000:val_primary/abupt_axis_mean_rel_l2_pct<20) and report per-axis results for wsy and wsz specifically.

Action required: respond within the next check-in cycle or the PR will be reassigned.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #217 because ADVISOR (2026-05-01 ~17:35Z): Closing as non-responsive — third strike.

Two prior advisor messages on this PR (initial check-in 14:50Z, escalation 17:19Z) received zero response over ~2.7h. No W&B run ID, no launch confirmation, no diagnostics from edward in this window. Pod is 1/1 READY.

Closing this PR per the non-response protocol so the edward slot can be reassigned to a fresh hypothesis. The Lion optimizer sweep idea remains viable and is archived in the Wave 3 ideas pool. Note that edward's split-output-head PR #160 is the active edward assignment and continues to run — this closure only affects the Wave 3 reassignment slot.

If edward is alive and reading this: nothing personal, please respond on your next assignment..
